Add support for secure Redis (TLS support)

qwerrrqw opened the pull request:
Add Redis TLS support for Heroku

Fixes #5500 

   base.py(286~287)
   - CELERY_BROKER_USE_SSL
   - CELERY_REDIS_BACKEND_USE_SSL

   production.py(51~54)
   - REDIS_SSL
   - Set to default=False, no impact on existing environments

   You only need to set REDIS_SSL=True in Heroku environment
   Add only new options without modifying existing settings

qwerrrqw commented:
## How to use Redis TLS in Heroku

1. Create Redis addon:
[bash] heroku addons:create heroku-redis:mini

2. Set environment variables:
[bash] heroku config:set REDIS_SSL=True
[bash] heroku config:set CELERY_BROKER_USE_SSL=True
[bash] heroku config:set CELERY_REDIS_BACKEND_USE_SSL=True


That's it! Your Redis connection will now use TLS encryption.

Note: These settings are optional and default to False, so they won't affect existing deployments unless explicitly enabled.

browniebroke commented:
We probably should update the documentation accordingly: https://cookiecutter-django.readthedocs.io/en/latest/3-deployment/deployment-on-heroku.html

The source is here: https://github.com/cookiecutter/cookiecutter-django/blob/master/docs/3-deployment/deployment-on-heroku.rst

qwerrrqw commented:
I've updated the documentation as requested:

deployment-on-heroku.rst 

1. Added Redis TLS configuration commands in the Script section: line 15-19

2. Added explanation in the Notes section under "Redis Configuration": line 71-74

Please let me know if any further changes are needed.

browniebroke commented on docs/3-deployment/deployment-on-heroku.rst:
```diff
@@ -58,6 +66,11 @@ Run these commands to deploy the project to Heroku:
 Notes
 -----
 
+Redis Configuration
++++++++++++++++++++
+
+Heroku Redis now requires TLS connections by default. The script above includes the necessary TLS configuration. These settings are optional and default to False, so they won't affect existing deployments unless explicitly enabled.
```
No need for this. While it seems like an important information right now, in month+ time, it will be irrelevant, so I'd rather not have this part in our docs

qwerrrqw commented:
I've made the requested changes:

1. Removed the duplicate Redis installation command
2. Moved the TLS configuration commands right after the Redis installation. line: 25-28
3. Removed the Redis Configuration section from the Notes

Thank you for your kind explanation

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/base.py:
```diff
@@ -283,6 +283,9 @@
 }
 
 REDIS_URL = env("REDIS_URL", default="redis://{% if cookiecutter.use_docker == 'y' %}redis{%else%}localhost{% endif %}:6379/0")
+CELERY_BROKER_USE_SSL = env.bool("CELERY_BROKER_USE_SSL", default=False)
+CELERY_REDIS_BACKEND_USE_SSL = env.bool("CELERY_REDIS_BACKEND_USE_SSL", default=False)
```
Maybe we can simplify things?

```suggestion
CELERY_BROKER_USE_SSL = env.bool("REDIS_SSL", default=False)
CELERY_REDIS_BACKEND_USE_SSL = env.bool("REDIS_SSL", default=False)
```

browniebroke commented on docs/3-deployment/deployment-on-heroku.rst:
```diff
@@ -14,13 +14,19 @@ Run these commands to deploy the project to Heroku:
 
+    # Enable Redis TLS support (required for new Heroku Redis instances)
+    heroku config:set REDIS_SSL=True
+    heroku config:set CELERY_BROKER_USE_SSL=True
+    heroku config:set CELERY_REDIS_BACKEND_USE_SSL=True
```
```suggestion
```

browniebroke reviewed (commented):
Sorry I missed that earlier, but feels like we have 3 settings to control the same thing, and they should all have the same value. Maybe let's put all under the same?

qwerrrqw commented:
I've simplified the Redis TLS configuration as advised. It's much better!

dqunbp commented:
These changes will not work, please check https://github.com/cookiecutter/cookiecutter-django/issues/5500

browniebroke commented:
> These changes will not work, please check #5500

Could you be a bit more specific as to why? Do you have any suggestions on how to improve the patch to make it work?

qwerrrqw commented:
I've attempted a PR based on the code provided in Issue #5500. Here are the issues encountered and their solutions:

1. **Ruff Rule Issue**
   - Using `import ssl` causes test failure in CI / pytest ubuntu-latest due to ruff rule violation
   - (Passes in macOS-latest and windows-latest)
   - Solution: Changed to use constant values directly instead of `import ssl`

2. **Environment Variable Handling**
   - Initial approach: Using `REDIS_TLS_URL` as a required environment variable
   - Problem: Multiple CI test failures (macOS-latest, Docker Basic, Docker Celery & DRF, Docker Webpack)
   - Solution: 
     - Maintain existing `REDIS_URL` configuration
     - Apply TLS settings only when `REDIS_SSL=True`

 **CI Execution Issue**
   - Problem: CI check stopped during execution...
   - Limitation: Cannot re-run workflow due to repository permissions
   - Question: How can I re-run the CI workflow?

qwerrrqw commented:
The latest changes from the master branch have been merged into fix/redis-TLS-support. After merging, the code is now compatible with the updated master branch, and all CL tests have passed successfully.

rristoferbowers commented:
> > These changes will not work, please check #5500
> 
> Could you be a bit more specific as to why? Do you have any suggestions on how to improve the patch to make it work?

I have tried these changes but it didn't work.
There is my final prod settings testes with Heroku
```python
REDIS_URL = env("REDIS_TLS_URL")
CELERY_REDIS_BACKEND_USE_SSL = {"ssl_cert_reqs": ssl.CERT_NONE}
CELERY_BROKER_USE_SSL = {"ssl_cert_reqs": ssl.CERT_NONE}
CELERY_BROKER_CONNECTION_RETRY_ON_STARTUP = True

# CACHES
# ------------------------------------------------------------------------------
CACHES = {
    "default": {
        "BACKEND": "django.core.cache.backends.redis.RedisCache",
        "LOCATION": REDIS_URL,
        "OPTIONS": {"ssl_cert_reqs": None},
    },
}
```

dqunbp commented:
> > These changes will not work, please check #5500
> 
> Could you be a bit more specific as to why? Do you have any suggestions on how to improve the patch to make it work?

@browniebroke 
I have tried these changes but it didn't work.
There is my final prod settings testes with Heroku
```python
REDIS_URL = env("REDIS_TLS_URL")
CELERY_REDIS_BACKEND_USE_SSL = {"ssl_cert_reqs": ssl.CERT_NONE}
CELERY_BROKER_USE_SSL = {"ssl_cert_reqs": ssl.CERT_NONE}
CELERY_BROKER_CONNECTION_RETRY_ON_STARTUP = True

# CACHES
# ------------------------------------------------------------------------------
CACHES = {
    "default": {
        "BACKEND": "django.core.cache.backends.redis.RedisCache",
        "LOCATION": REDIS_URL,
        "OPTIONS": {"ssl_cert_reqs": None},
    },
}
```

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/production.py:
```diff
@@ -48,6 +48,20 @@
+if REDIS_SSL:
+    CELERY_REDIS_BACKEND_USE_SSL = {"ssl_cert_reqs": CERT_NONE}
+    CELERY_BROKER_USE_SSL = {"ssl_cert_reqs": CERT_NONE}
+    CACHES["default"]["OPTIONS"]["CONNECTION_POOL_CLASS"] = "redis.connection.SSLConnection"
+    CACHES["default"]["OPTIONS"]["SSL_CERT_REQS"] = CERT_NONE
```
Why do you need that? The cache seems to work fine on my end with the new secure redis connection string... What am I missing?

qwerrrqw replied:
To facilitate testing in a development environment, we have added a setting to disable SSL certificate verification.
However, as you pointed out, if it works well with the default settings,
I think it would be cleaner to remove unnecessary settings and handle SSL connections with only the 'rediss://' URL.

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/production.py:
```diff
@@ -48,6 +48,20 @@
 
+
+REDIS_URL = env("REDIS_TLS_URL", default=env("REDIS_URL"))
+REDIS_SSL = env.bool("REDIS_SSL", default=False)
```
This should probably be in `base.py` and can be derived from the `REDIS_URL`, I think

```suggestion
REDIS_SSL = REDIS_URL.startswith("rediss://")
```

browniebroke commented on docs/3-deployment/deployment-on-heroku.rst:
```diff
@@ -14,13 +14,17 @@ Run these commands to deploy the project to Heroku:
     heroku addons:create heroku-redis:mini
 
+    # Enable Redis TLS support (required for new Heroku Redis instances)
+    heroku config:set REDIS_SSL=True
```
No need for that update if we derive it from the redis URL

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/production.py:
```diff
@@ -48,6 +48,20 @@
+CERT_NONE = 0
+
+if REDIS_SSL:
+    CELERY_REDIS_BACKEND_USE_SSL = {"ssl_cert_reqs": CERT_NONE}
+    CELERY_BROKER_USE_SSL = {"ssl_cert_reqs": CERT_NONE}
```
Can you move these next to the other celery settings in `base.py`? Try to keep backend and broker related settings together please and link to the celery docs for each. That should give something like this:


```diff
  # Celery
  # ------------------------------------------------------------------------------
  if USE_TZ:
      # https://docs.celeryq.dev/en/stable/userguide/configuration.html#std:setting-timezone
      CELERY_TIMEZONE = TIME_ZONE
  # https://docs.celeryq.dev/en/stable/userguide/configuration.html#std:setting-broker_url
  CELERY_BROKER_URL = REDIS_URL
+ # https://docs.celeryq.dev/en/stable/userguide/configuration.html#redis-backend-use-ssl
+ CELERY_BROKER_USE_SSL = {"ssl_cert_reqs": ssl.CERT_NONE} if REDIS_SSL else None
  # https://docs.celeryq.dev/en/stable/userguide/configuration.html#std:setting-result_backend
  CELERY_RESULT_BACKEND = REDIS_URL
+ # https://docs.celeryq.dev/en/stable/userguide/configuration.html#redis-backend-use-ssl
+ CELERY_REDIS_BACKEND_USE_SSL = CELERY_BROKER_USE_SSL
  # https://docs.celeryq.dev/en/stable/userguide/configuration.html#result-extended
  CELERY_RESULT_EXTENDED = True
```

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/production.py:
```diff
@@ -48,6 +48,20 @@
+    CACHES["default"]["OPTIONS"]["CONNECTION_POOL_CLASS"] = "redis.connection.SSLConnection"
+    CACHES["default"]["OPTIONS"]["SSL_CERT_REQS"] = CERT_NONE
+
+CELERY_BROKER_CONNECTION_RETRY_ON_STARTUP = True
```
That's out of scope I think, please revert, see https://github.com/cookiecutter/cookiecutter-django/issues/5085

qwerrrqw replied:
Yes, the CELERY_BROKER_CONNECTION_RETRY_ON_STARTUP setting is better covered in issue #5085, I will remove it from this PR.

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/base.py:
```diff
@@ -1,10 +1,14 @@
+# ruff: noqa: ERA001, E501, E402
 """Base settings to build other settings files upon."""
 
 from pathlib import Path
 
 import environ
 
+{% if cookiecutter.use_celery == 'y' %}
+import ssl  # noqa: E402
+{% endif %}
```
```suggestion
{% if cookiecutter.use_celery == 'y' -%}
import ssl
{%- endif %}
from pathlib import Path

import environ

```

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/base.py:
```diff
@@ -1,10 +1,14 @@
-# ruff: noqa: ERA001, E501
+# ruff: noqa: ERA001, E501, E402
```
```suggestion
# ruff: noqa: ERA001, E501
```

browniebroke commented on {{cookiecutter.project_slug}}/config/settings/base.py:
```diff
@@ -1,10 +1,14 @@
 
 import environ
 
+
```
```suggestion
```

browniebroke commented:
Thanks for the contribution 🎉
